pkg/prometheus/statefulset.go

```diff
@@ -1076,3 +676,488 @@ func queryLogFilePath(queryLogFile string) string {
+		})
+		promVolumeMounts = append(promVolumeMounts, v1.VolumeMount{
+			Name:      name,
+			ReadOnly:  true,
+			MountPath: configmapsDir + c,
+		})
+	}
+
+	return volumes, promVolumeMounts, nil
+}
+
+// addServerVolumes adds volumes to the statefulset spec that are specific to Prometheus Server
+func addServerVolumes(volumes *[]v1.Volume, volumeMounts *[]v1.VolumeMount, queryLogFile string, ruleConfigMapNames []string) {
+	if volume, ok := queryLogFileVolume(queryLogFile); ok {
+		*volumes = append(*volumes, volume)
+	}
+
+	for _, name := range ruleConfigMapNames {
+		*volumes = append(*volumes, v1.Volume{
+			Name: name,
+			VolumeSource: v1.VolumeSource{
+				ConfigMap: &v1.ConfigMapVolumeSource{
+					LocalObjectReference: v1.LocalObjectReference{
+						Name: name,
+					},
+				},
+			},
+		})
+	}
+
+	for _, name := range ruleConfigMapNames {
+		*volumeMounts = append(*volumeMounts, v1.VolumeMount{
+			Name:      name,
+			MountPath: rulesDir + "/" + name,
+		})
+	}
+
+	if vmount, ok := queryLogFileVolumeMount(queryLogFile); ok {
+		*volumeMounts = append(*volumeMounts, vmount)
+	}
+}
+
+func probeHandler(probePath string, cpf monitoringv1.CommonPrometheusFields, webConfigGenerator *ConfigGenerator, webRoutePrefix string) v1.ProbeHandler {
+	probePath = path.Clean(webRoutePrefix + probePath)
+	handler := v1.ProbeHandler{}
+	if cpf.ListenLocal {
+		probeURL := url.URL{
+			Scheme: "http",
+			Host:   "localhost:9090",
+			Path:   probePath,
+		}
+		handler.Exec = &v1.ExecAction{
+			Command: []string{
+				"sh",
+				"-c",
+				fmt.Sprintf(
+					`if [ -x "$(command -v curl)" ]; then exec %s; elif [ -x "$(command -v wget)" ]; then exec %s; else exit 1; fi`,
+					operator.CurlProber(probeURL.String()),
+					operator.WgetProber(probeURL.String()),
+				),
+			},
+		}
+		return handler
+	}
+
+	handler.HTTPGet = &v1.HTTPGetAction{
+		Path: probePath,
+		Port: intstr.FromString(cpf.PortName),
+	}
+	if cpf.Web != nil && cpf.Web.TLSConfig != nil && webConfigGenerator.IsCompatible() {
+		handler.HTTPGet.Scheme = v1.URISchemeHTTPS
+	}
+	return handler
+}
+
+func buildPodMetadata(cpf monitoringv1.CommonPrometheusFields, cg *ConfigGenerator) (map[string]string, map[string]string) {
+	podAnnotations := map[string]string{
+		"kubectl.kubernetes.io/default-container": "prometheus",
+	}
+	podLabels := map[string]string{
+		"app.kubernetes.io/version": cg.version.String(),
+	}
+
+	if cpf.PodMetadata != nil {
+		if cpf.PodMetadata.Labels != nil {
+			for k, v := range cpf.PodMetadata.Labels {
+				podLabels[k] = v
+			}
+		}
+		if cpf.PodMetadata.Annotations != nil {
+			for k, v := range cpf.PodMetadata.Annotations {
+				podAnnotations[k] = v
+			}
+		}
+	}
+
+	return podAnnotations, podLabels
+}
+
+func configureThanos(
+	additionalContainers *[]v1.Container,
+	disableCompaction *bool,
+	cpf monitoringv1.CommonPrometheusFields,
+	thanos *monitoringv1.ThanosSpec,
+	c *operator.Config,
+	prometheusURIScheme, webRoutePrefix, volName string) error {
```

Comment:
suggestion: the caller can append the container to the slice.
```suggestion
func createThanosContainer(
	disableCompaction *bool,
	cpf monitoringv1.CommonPrometheusFields,
	thanos *monitoringv1.ThanosSpec,
	c *operator.Config,
	prometheusURIScheme, webRoutePrefix, volName string) ([]v1.Container, error) {
```